File: apps/server/app/api/utils/dom_parser/filters.py
import logging

from bs4 import NavigableString, Tag

logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger("DOM-Analyzer")


unaccepted_leaf_element_tags_set = {
    "script", "style", "meta", "link", "iframe", "svg", "canvas", "video", "audio"
}


def tag_wise_filter(element_tag: str) -> bool:
    """Filter elements based on their tag name."""
    try:
        if not element_tag:
            logger.debug("Empty element tag provided to tag_wise_filter")
            return False

        normalized_tag = element_tag.lower()
        logger.debug(f"Checking tag filter for: {normalized_tag}")

        if normalized_tag in unaccepted_leaf_element_tags_set:
            logger.debug(f"Tag {normalized_tag} is in unaccepted set")
            return False

        logger.debug(f"Tag {normalized_tag} is accepted")
        return True
    except Exception as error:
        logger.error(f"Error in tag_wise_filter: {error}")
        return False  # Fail safe - reject elements that cause errors


def is_element_visible(element: Tag) -> bool:
    """
    Estimate if an element is likely visible based on HTML attributes.
    This is a simplified version since we don't have access to rendered styles in Python.
    """
    try:
        if not element:
            logger.debug("Null element provided to is_element_visible")
            return False

        tag_name = element.name.lower() if element.name else ""
        logger.debug(f"Checking visibility for {tag_name}")

        # Check style attribute for display:none or visibility:hidden
        style_attr = element.get('style', '')
        if 'display:none' in style_attr.lower() or 'visibility:hidden' in style_attr.lower():
            return False

        # Check for common hidden attributes
        if element.get('hidden') is not None or element.get('aria-hidden') == 'true':
            return False

        # Check class names for common patterns indicating hidden elements
        class_attr = element.get('class', [])
        class_names = class_attr if isinstance(
            class_attr, list) else str(class_attr).split()
        hidden_classes = ['hidden', 'invisible', 'displaynone', 'nodisplay']
        if any(hidden_class in class_name.lower() for class_name in class_names for hidden_class in hidden_classes):
            return False

        logger.debug(f"Element visibility result: True")
        return True
    except Exception as error:
        logger.error(f"Error in is_element_visible: {error}")
        return False  # Fail safe - consider problematic elements as not visible


def is_top_element(element: Tag) -> bool:
    """
    Simplified version of isTopElement since we can't determine stacking in static HTML.
    Assumes elements are top unless they have specific attributes suggesting otherwise.
    """
    try:
        if not element:
            logger.debug("Null element provided to is_top_element")
            return False

        tag_name = element.name.lower() if element.name else ""
        logger.debug(f"Checking if {tag_name} is top element")

        # Check for elements that are likely to be overlaid by others
        if tag_name in ['body', 'html']:
            return False

        # Check for background elements
        style_attr = element.get('style', '').lower()
        if 'z-index:-' in style_attr or 'z-index: -' in style_attr:
            return False

        return True
    except Exception as error:
        logger.error(f"Error in is_top_element: {error}")
        return True  # Default to visible in case of errors


def is_text_node_visible(text_node: NavigableString) -> bool:
    """Check if a text node is likely visible."""
    try:
        if not text_node or not text_node.strip():
            logger.debug("Null or empty text node provided")
            return False

        logger.debug("Checking text node visibility")

        # Check if parent element is visible
        parent = text_node.parent
        if not parent or not is_element_visible(parent):
            return False

        return True
    except Exception as error:
        logger.error(f"Error in is_text_node_visible: {error}")
        return False  # Fail safe - consider problematic text nodes as not visible


def is_interactive_element(element: Tag) -> bool:
    """Check if an element is interactive based on its tag and attributes."""
    try:
        if not element:
            logger.debug("Null element provided to is_interactive_element")
            return False

        tag_name = element.name.lower() if element.name else ""
        logger.debug(f"Checking if {tag_name} is interactive")

        # Immediately return false for body tag
        if tag_name == "body":
            logger.debug("Body tag is not interactive")
            return False

        # Base interactive elements and roles
        interactive_elements = {
            "a", "button", "details", "embed", "input", "label",
            "menu", "menuitem", "object", "select", "textarea", "summary"
        }

        interactive_roles = {
            "button", "menu", "menuitem", "link", "checkbox", "radio",
            "slider", "tab", "tabpanel", "textbox", "combobox", "grid",
            "listbox", "option", "progressbar", "scrollbar", "searchbox",
            "switch", "tree", "treeitem", "spinbutton", "tooltip",
            "a-button-inner", "a-dropdown-button", "click", "menuitemcheckbox",
            "menuitemradio", "a-button-text", "button-text", "button-icon",
            "button-icon-only", "button-text-icon-only", "dropdown", "combobox"
        }

        role = element.get('role', '')
        aria_role = element.get('aria-role', '')
        tab_index = element.get('tabindex')

        # Check for specific class
        classes = element.get('class', [])
        class_list = classes if isinstance(
            classes, list) else str(classes).split()
        has_address_input_class = "address-input__container__input" in class_list

        # Basic role/attribute checks
        parent_tag = element.parent.name.lower(
        ) if element.parent and element.parent.name else ""
        has_interactive_role = (
            has_address_input_class or
            tag_name in interactive_elements or
            role in interactive_roles or
            aria_role in interactive_roles or
            (tab_index is not None and tab_index != "-1" and parent_tag != "body") or
            element.get('data-action') == "a-dropdown-select" or
            element.get('data-action') == "a-dropdown-button"
        )

        if has_interactive_role:
            logger.debug(f"Element has interactive role: {tag_name}")
            return True

        # Check for event listeners
        has_click_handler = (
            element.get('onclick') is not None or
            element.has_attr('ng-click') or
            element.has_attr('@click') or
            element.has_attr('v-on:click')
        )

        # Check for ARIA properties that suggest interactivity
        has_aria_props = (
            element.has_attr('aria-expanded') or
            element.has_attr('aria-pressed') or
            element.has_attr('aria-selected') or
            element.has_attr('aria-checked')
        )

        # Check if element is draggable
        is_draggable = element.get('draggable') == "true"

        # Additional check to prevent body from being marked as interactive
        if tag_name == "body" or parent_tag == "body":
            logger.debug(
                "Element is body or direct child of body - not interactive")
            return False

        is_interactive = (
            has_aria_props or
            has_click_handler or
            is_draggable
        )

        logger.debug(
            f"Element interactivity: {is_interactive} (ariaProps: {has_aria_props}, clickHandler: {has_click_handler}, draggable: {is_draggable})")
        return is_interactive
    except Exception as error:
        logger.error(f"Error in is_interactive_element: {error}")
        return False  # Fail safe - consider problematic elements as not interactive

File: packages/core/src/utils/filters.ts
// Logger helper
const logger = {
  debug: (message: string, ...data: any[]) =>
    console.debug(`[Filters] ${message}`, ...data),
  error: (message: string, error?: any) =>
    console.error(`[Filters] ${message}`, error),
};

export const unAcceptedLeafElementTagsSet = new Set<string>([
  "script",
  "style",
  "meta",
  "link",
  "iframe",
  "svg",
  "canvas",
  "video",
  "audio",
]);

export const tagWiseFilter = (elementTag: string) => {
  try {
    if (!elementTag) {
      logger.debug("Empty element tag provided to tagWiseFilter");
      return false;
    }

    const normalizedTag = elementTag.toLowerCase();
    logger.debug(`Checking tag filter for: ${normalizedTag}`);

    if (unAcceptedLeafElementTagsSet.has(normalizedTag)) {
      logger.debug(`Tag ${normalizedTag} is in unaccepted set`);
      return false;
    }

    logger.debug(`Tag ${normalizedTag} is accepted`);
    return true;
  } catch (error) {
    logger.error("Error in tagWiseFilter", error);
    return false; // Fail safe - reject elements that cause errors
  }
};

export function isElementVisible(element: Element): boolean {
  try {
    if (!element) {
      logger.debug("Null element provided to isElementVisible");
      return false;
    }

    logger.debug(`Checking visibility for ${element.tagName.toLowerCase()}`);

    if (!window) {
      logger.error("Window object not available");
      return false;
    }

    const style = window.getComputedStyle(element);

    const isVisible =
      element.clientWidth > 0 &&
      element.clientHeight > 0 &&
      style.visibility !== "hidden" &&
      style.display !== "none";

    logger.debug(`Element visibility result: ${isVisible}`);
    return isVisible;
  } catch (error) {
    logger.error("Error in isElementVisible", error);
    return false; // Fail safe - consider problematic elements as not visible
  }
}

/**
 * Checks if an element is the top element at its position.
 */
export const isTopElement = (element: Element) => {
  try {
    if (!element) {
      logger.debug("Null element provided to isTopElement");
      return false;
    }

    logger.debug(`Checking if ${element.tagName.toLowerCase()} is top element`);

    // Find the correct document context and root element
    let doc = element.ownerDocument;
    if (!doc) {
      logger.error("Unable to find owner document");
      return false;
    }

    // If we're in an iframe, elements are considered top by default
    if (doc !== window.document) {
      logger.debug("Element is in iframe - considering as top element");
      return true;
    }

    // For shadow DOM, we need to check within its own root context
    const shadowRoot = element.getRootNode();
    if (shadowRoot instanceof ShadowRoot) {
      logger.debug("Element is in shadow DOM");

      try {
        const rect = element.getBoundingClientRect();
        const point = {
          x: rect.left + rect.width / 2,
          y: rect.top + rect.height / 2,
        };

        logger.debug(`Checking point (${point.x}, ${point.y}) in shadow root`);

        // Use shadow root's elementFromPoint to check within shadow DOM context
        const topEl = shadowRoot.elementFromPoint(point.x, point.y);
        if (!topEl) {
          logger.debug("No element found at point in shadow DOM");
          return false;
        }

        // Check if the element or any of its parents match our target element
        let current: Element | null = topEl;
        while (current && !(current instanceof ShadowRoot)) {
          if (current === element) {
            logger.debug("Element is top element in shadow DOM");
            return true;
          }
          current = current.parentElement;
        }

        logger.debug("Element is not top element in shadow DOM");
        return false;
      } catch (e) {
        logger.error("Error checking element in shadow DOM", e);
        return true; // If we can't determine, consider it visible
      }
    }

    // Regular DOM elements
    try {
      const rect = element.getBoundingClientRect();
      const viewportExpansion = 0;

      // Calculate expanded viewport boundaries including scroll position
      const scrollX = window.scrollX;
      const scrollY = window.scrollY;
      const viewportTop = -viewportExpansion + scrollY;
      const viewportLeft = -viewportExpansion + scrollX;
      const viewportBottom = window.innerHeight + viewportExpansion + scrollY;
      const viewportRight = window.innerWidth + viewportExpansion + scrollX;

      // Get absolute element position
      const absTop = rect.top + scrollY;
      const absLeft = rect.left + scrollX;
      const absBottom = rect.bottom + scrollY;
      const absRight = rect.right + scrollX;

      // Skip if element is completely outside expanded viewport
      if (
        absBottom < viewportTop ||
        absTop > viewportBottom ||
        absRight < viewportLeft ||
        absLeft > viewportRight
      ) {
        logger.debug("Element is outside viewport");
        return false;
      }

      // For elements within expanded viewport, check if they're the top element
      const centerX = rect.left + rect.width / 2;
      const centerY = rect.top + rect.height / 2;

      // Only clamp the point if it's outside the actual document
      const point = {
        x: centerX,
        y: centerY,
      };

      logger.debug(`Checking point (${point.x}, ${point.y}) in document`);

      if (
        point.x < 0 ||
        point.x >= window.innerWidth ||
        point.y < 0 ||
        point.y >= window.innerHeight
      ) {
        logger.debug("Element center is outside viewport - considering as top");
        return true; // Consider elements with center outside viewport as visible
      }

      const topEl = document.elementFromPoint(point.x, point.y);
      if (!topEl) {
        logger.debug("No element found at point");
        return false;
      }

      let current = topEl;
      while (current && current !== document.documentElement) {
        if (current === element) {
          logger.debug("Element is top element");
          return true;
        }
        current = current.parentElement as Element;
      }

      logger.debug("Element is not top element");
      return false;
    } catch (e) {
      logger.error("Error checking if regular DOM element is top", e);
      return true; // If we can't determine, consider it visible
    }
  } catch (error) {
    logger.error("Fatal error in isTopElement", error);
    return true; // Default to visible in case of errors
  }
};

/**
 * Checks if a text node is visible.
 */
export const isTextNodeVisible = (textNode: Text): boolean => {
  try {
    if (!textNode) {
      logger.debug("Null text node provided");
      return false;
    }

    logger.debug("Checking text node visibility");

    if (!document) {
      logger.error("Document object not available");
      return false;
    }

    // Create a range and get its bounding rect
    const range = document.createRange();
    range.selectNodeContents(textNode);
    const rect = range.getBoundingClientRect();

    if (!rect || !textNode.parentElement) {
      logger.debug("No bounding rect or parent element for text node");
      return false;
    }

    // Check visibility based on several factors
    const hasSize = rect.width !== 0 && rect.height !== 0;
    const isInViewport = rect.top >= 0 && rect.top <= window.innerHeight;

    let parentVisible = false;
    try {
      parentVisible = textNode.parentElement.checkVisibility({
        checkOpacity: true,
        checkVisibilityCSS: true,
      });
    } catch (e) {
      logger.error("Error checking parent visibility", e);
      // Fall back to basic parent check if checkVisibility isn't supported
      const parentStyle = window.getComputedStyle(textNode.parentElement);
      parentVisible =
        parentStyle.display !== "none" &&
        parentStyle.visibility !== "hidden" &&
        parentStyle.opacity !== "0";
    }

    const isVisible = hasSize && isInViewport && parentVisible;

    logger.debug(
      `Text node visibility: ${isVisible} (hasSize: ${hasSize}, inViewport: ${isInViewport}, parentVisible: ${parentVisible})`
    );
    return isVisible;
  } catch (error) {
    logger.error("Error in isTextNodeVisible", error);
    return false; // Fail safe - consider problematic text nodes as not visible
  }
};

// Referred from browser-use (https://github.com/browser-use/browser-use)
export const isInteractiveElement = (element: Element) => {
  try {
    if (!element) {
      logger.debug("Null element provided to isInteractiveElement");
      return false;
    }

    const tagName = element.tagName.toLowerCase();
    logger.debug(`Checking if ${tagName} is interactive`);

    // Immediately return false for body tag
    if (tagName === "body") {
      logger.debug("Body tag is not interactive");
      return false;
    }

    // Base interactive elements and roles
    const interactiveElements = new Set([
      "a",
      "button",
      "details",
      "embed",
      "input",
      "label",
      "menu",
      "menuitem",
      "object",
      "select",
      "textarea",
      "summary",
    ]);

    const interactiveRoles = new Set([
      "button",
      "menu",
      "menuitem",
      "link",
      "checkbox",
      "radio",
      "slider",
      "tab",
      "tabpanel",
      "textbox",
      "combobox",
      "grid",
      "listbox",
      "option",
      "progressbar",
      "scrollbar",
      "searchbox",
      "switch",
      "tree",
      "treeitem",
      "spinbutton",
      "tooltip",
      "a-button-inner",
      "a-dropdown-button",
      "click",
      "menuitemcheckbox",
      "menuitemradio",
      "a-button-text",
      "button-text",
      "button-icon",
      "button-icon-only",
      "button-text-icon-only",
      "dropdown",
      "combobox",
    ]);

    const role = element.getAttribute("role");
    const ariaRole = element.getAttribute("aria-role");
    const tabIndex = element.getAttribute("tabindex");

    // Add check for specific class
    const hasAddressInputClass = element.classList.contains(
      "address-input__container__input"
    );

    // Basic role/attribute checks
    const hasInteractiveRole =
      hasAddressInputClass ||
      interactiveElements.has(tagName) ||
      interactiveRoles.has(role || "") ||
      interactiveRoles.has(ariaRole || "") ||
      (tabIndex !== null &&
        tabIndex !== "-1" &&
        element.parentElement?.tagName.toLowerCase() !== "body") ||
      element.getAttribute("data-action") === "a-dropdown-select" ||
      element.getAttribute("data-action") === "a-dropdown-button";

    if (hasInteractiveRole) {
      logger.debug(`Element has interactive role: ${tagName}`);
      return true;
    }

    // Get computed style
    let style;
    try {
      style = window.getComputedStyle(element);
    } catch (styleError) {
      logger.error("Error getting computed style", styleError);
      // Continue with partial checks if style can't be retrieved
    }

    // Check for event listeners
    const hasClickHandler =
      element.getAttribute("onclick") !== null ||
      element.hasAttribute("ng-click") ||
      element.hasAttribute("@click") ||
      element.hasAttribute("v-on:click");

    // Helper function to safely get event listeners
    function getEventListeners(el: Element) {
      try {
        // Try to get listeners using Chrome DevTools API
        return (window as any).getEventListeners?.(el) || {};
      } catch (e) {
        logger.debug("getEventListeners not available, using fallback");
        // Fallback: check for common event properties
        const listeners = {};

        // List of common event types to check
        const eventTypes = [
          "click",
          "mousedown",
          "mouseup",
          "touchstart",
          "touchend",
          "keydown",
          "keyup",
          "focus",
          "blur",
        ];

        for (const type of eventTypes) {
          const handlerProp = `on${type}` as keyof Element;
          const handler = el[handlerProp];
          if (handler) {
            (listeners as Record<string, any>)[type] = [
              {
                listener: handler,
                useCapture: false,
              },
            ];
          }
        }

        return listeners as Record<
          string,
          Array<{ listener: any; useCapture: boolean }>
        >;
      }
    }

    // Check for click-related events on the element itself
    let listeners: Record<
      string,
      Array<{ listener: any; useCapture: boolean }> | undefined
    > = {};
    try {
      listeners = getEventListeners(element);
    } catch (listenerError) {
      logger.error("Error getting event listeners", listenerError);
      // Continue with partial checks if listeners can't be retrieved
    }

    const hasClickListeners =
      (typeof listeners["click"] !== "undefined" &&
        listeners["click"]?.length > 0) ||
      (typeof listeners["mousedown"] !== "undefined" &&
        listeners["mousedown"]?.length > 0) ||
      (typeof listeners["mouseup"] !== "undefined" &&
        listeners["mouseup"]?.length > 0) ||
      (typeof listeners["touchstart"] !== "undefined" &&
        listeners["touchstart"]?.length > 0) ||
      (typeof listeners["touchend"] !== "undefined" &&
        listeners["touchend"]?.length > 0);

    // Check for ARIA properties that suggest interactivity
    const hasAriaProps =
      element.hasAttribute("aria-expanded") ||
      element.hasAttribute("aria-pressed") ||
      element.hasAttribute("aria-selected") ||
      element.hasAttribute("aria-checked");

    // Check for form-related functionality
    const isFormRelated =
      element.hasAttribute("contenteditable") ||
      (style && style.userSelect !== "none");

    // Check if element is draggable
    const isDraggable = element.getAttribute("draggable") === "true";

    // Additional check to prevent body from being marked as interactive
    if (
      tagName === "body" ||
      element.parentElement?.tagName.toLowerCase() === "body"
    ) {
      logger.debug("Element is body or direct child of body - not interactive");
      return false;
    }

    const isInteractive =
      hasAriaProps || hasClickHandler || hasClickListeners || isDraggable;

    logger.debug(
      `Element interactivity: ${isInteractive} (ariaProps: ${hasAriaProps}, clickHandler: ${hasClickHandler}, clickListeners: ${hasClickListeners}, draggable: ${isDraggable})`
    );
    return isInteractive;
  } catch (error) {
    logger.error("Error in isInteractiveElement", error);
    return false; // Fail safe - consider problematic elements as not interactive
  }
};
